src/FieldWarning/Assets/Units/Component/Movement/VehicleMovementStrategy.cs

```diff
@@ -0,0 +1,308 @@
+        public Vector3 NextRotation;
+        public float FinalHeading;
+
+        public VehicleMovementStrategy(
+                DataComponent data, TerrainMap map, Transform transform, MobilityType mobility)
+        {
+            _data = data;
+            _terrainMap = map;
+            _transform = transform;
+            _mobility = mobility;
+        }
+
+        public void PlanMovement()
+        {
+            float distanceToWaypoint = 
+                    Pathfinder.HasDestination() ? 
+                            (Pathfinder.GetWaypoint() - NextPosition).magnitude : 0f;
+            float linSpeed = Mathf.Abs(_linVelocity);
+            float rotationSpeed = CalculateRotationSpeed(linSpeed);
+
+            //float remainingTurn = CalculateRemainingTurn(targetHeading);
+            float targetHeading = GetTargetHeading();
+            float turnForward = 0f;
+            float turnReverse = 0f;
+            if (targetHeading != NO_HEADING) {
+                turnForward = (targetHeading - NextRotation.y - Mathf.PI / 2).unwrapRadian();
+                turnReverse = (targetHeading - NextRotation.y + Mathf.PI / 2).unwrapRadian();
+            }
+            bool isReverse = ShouldReverse(
+                    _linVelocity, distanceToWaypoint, rotationSpeed, turnForward, turnReverse);
+            float remainingTurn = isReverse ? turnReverse : turnForward;
+
+            float targetSpeed = CalculateTargetSpeed(
+                    distanceToWaypoint, remainingTurn, linSpeed, rotationSpeed);
+            if (isReverse)
+                targetSpeed = -targetSpeed;
+
+            DoRotationalMotion(remainingTurn, rotationSpeed);
+            DoLinearMotion(targetSpeed);
+        }
+
+        public void UpdateMapOrientation(Vector3 forward, Vector3 right)
+        {
+            int terrainType = _terrainMap == null ? 
+                    TerrainMap.PLAIN : _terrainMap.GetTerrainType(_transform.position);
+            if (terrainType == TerrainMap.BRIDGE)
+            {
+                _terrainTiltForward = 0f;
+                _terrainTiltRight = 0f;
+                _terrainHeight = TerrainMap.BRIDGE_HEIGHT;
+            }
+            else if (terrainType == TerrainMap.WATER)
+            {
+                _terrainTiltForward = 0f;
+                _terrainTiltRight = 0f;
+                _terrainHeight = _terrainMap.WATER_HEIGHT;
+            }
+            else
+            {
+                // This way of doing the rotation should look nice because the unit 
+                // won't sink into the ground much assuming length and width are set 
+                // correctly, but it is not very fast.
+
+                Terrain terrain = _terrainMap.GetTerrainAtPos(_transform.position);
+                if (terrain == null)
+                    return;
+
+                // Apparently our forward and backward are opposite of the Unity convention
+                float frontHeight = terrain.SampleHeight(
```

Comment:
so all this sampleheight stuff we do is extremely slow.. for one or two things its fine.. but its better if we cache the sample height and use that cached data.. just FYI